pkg/tfbridge/provider.go

```diff
@@ -59,6 +59,106 @@ type Resource struct {
 	TFName string           // the Terraform resource name.
 }
 
+// collectImporterAttributes runs the Terraform Importer defined on the Resource for the given
+// resource ID, and adds any non-ID attributes to the inputs map.
+func (res *Resource) collectImporterAttributes(resourceID resource.ID, inputs map[string]string) error {
+	// There is nothing to do here if the resource doesn't have an importer defined in the
+	// Terraform schema.
+	if res.TF.Importer == nil {
+		return nil
+	}
+
+	glog.V(9).Infof("%s has TF Importer", res.TFName)
+
+	id := resourceID.String()
+
+	// Prepare a Terraform ResourceData for the importer
+	data := res.TF.Data(nil)
+	data.SetId(id)
+	data.SetType(res.TFName)
+
+	// Run the importer defined in the Terraform resource schema
+	results, err := res.TF.Importer.State(data, nil)
+	if err != nil {
+		return errors.Wrapf(err, "importing %s", id)
+	}
+
+	// No resources were returned - error out
+	if len(results) < 1 {
+		return errors.Errorf("importer for %s returned no resources", id)
+	}
```

Comment:
Nit - and it may be cleaner the way you have it - but I believe this condition is redundant with the later check that returns the same error.

Reply:
It is redundant (and vestigial of the previous pass at this pull request). I’m not sure it’s necessarily worse, though these are two different error conditions:

- Here, the Terraform importer returned no results at all
- In the repeated check later, there may have been several results from the Terraform importer but none satisfied our criteria for matching the imported resource.

Depending on whether we want to watch out explicitly for that condition, it might be better to make the error message reflect the different nature of the failure to make identifying any issues which come up in the wild a bit easier? If we don’t care about distinguishing between these conditions, we can remove this check as you suggest.

Reply:
In the latest version of this I’ve opted to keep the additional check, but change the wording of the later path to indicate that there were no matching results rather than no results returned.